examples/do_MLICA.py

```diff
@@ -0,0 +1,134 @@
+
+# 50HZ and 60HZ notch filter to remove noise
+raw = noise_reducer(raw_fname, raw=raw_nr, refnotch=refnotch, return_raw=True)
+
+picks = mne.pick_types(raw.info, meg=True, eeg=False, eog=False,
+                       stim=False, exclude='bads')
+
+raw.filter(0., 45., picks=picks, filter_length='auto',
+           l_trans_bandwidth='auto', h_trans_bandwidth='auto',
+           n_jobs=njobs, method='fir', phase='zero',
+           fir_window='hamming')
+
+# downsample the data to 250 Hz, necessary for the model
+raw_ds = raw.copy().resample(250, npad='auto', window='boxcar', stim_picks=None, n_jobs=njobs, events=None)
+raw_ds_chop = raw_ds.copy().crop(tmin=tmin*4./1000, tmax=tmax*4./1000)  # downsampled raw
+raw_chop = raw.copy().crop(tmin=tmin*4./1000, tmax=tmax*4./1000)
+
+ica = ICA(method='fastica', n_components=n_components, random_state=None,
+          max_pca_components=None, max_iter=1000, verbose=None)
+
+# do the ICA decomposition on downsampled raw
+ica.fit(raw_ds_chop, picks=picks, reject={'mag': 5e-12},
+        verbose=None)
+
+sources = ica.get_sources(raw_ds_chop)._data
+
+# extract temporal and spatial components
+mm = np.float32(np.dot(ica.mixing_matrix_[:, :].T,
+                       ica.pca_components_[:ica.n_components_]))
+
+# use [:, :15000] to make sure it's 15000 data points
+chop = sources[:, :15000]
+chop_reshaped = np.reshape(chop, (len(chop), len(chop[0]), 1))
+
+model_scores = model.predict([mm, chop_reshaped], verbose=1)
+
+bads_MLICA = []
+
+print model_scores
+
+for idx in range(0, len(model_scores)):
+    if model_scores[idx][0] > MLICA_threshold:
+        bads_MLICA.append(idx)
+
+ica.exclude = bads_MLICA
+
+# visualisation
+ica.plot_sources(raw_ds_chop, block=True)
+
+# compare MLICA to results from correlation and ctps analysis
+ica.exclude = []
+
+print 'Computing scores and identifying components..'
+ecg_scores = ica.score_sources(raw_chop, target=ecg_ch, score_func='pearsonr',
+                               l_freq=flow_ecg, h_freq=fhigh_ecg, verbose=False)
+# horizontal channel
+eog1_scores = ica.score_sources(raw_chop, target=eog1_ch, score_func='pearsonr',
+                                l_freq=flow_eog, h_freq=fhigh_eog, verbose=False)
+# vertical channel
+eog2_scores = ica.score_sources(raw_chop, target=eog2_ch, score_func='pearsonr',
+                                l_freq=flow_eog, h_freq=fhigh_eog, verbose=False)
+
+# print the top ecg, eog correlation scores
+ecg_inds = np.where(np.abs(ecg_scores) > ecg_thresh)[0]
+eog1_inds = np.where(np.abs(eog1_scores) > eog_thresh)[0]
+eog2_inds = np.where(np.abs(eog2_scores) > eog_thresh)[0]
+highly_corr = list(set(np.concatenate((ecg_inds, eog1_inds, eog2_inds))))
+print 'Highly correlated artifact components are', highly_corr
+
+
+bads_corr_ctps = []
+bads_corr_ctps += highly_corr
+
+# get ECG/EOG related components using JuMEG
+ic_ecg = get_ics_cardiac(raw_chop, ica,
```

Comment:
why are we computing ecg/eog scores and identifying components twice? Do the results vary between the method above and the get_ics_cardiac/ocular methods below?

Reply:
We could also break it up into three parts: MLICA, correlation using mne, and ctps using jumeg.

The reason why I combined both is that with other files it happened a couple of times that the results were different. Here it probably won't make a difference since there are only three components anyways. What would you recommend in general?

Reply:
I'd say we keep it simple and do only get_ics_cardiac/ocular, ecg scoring with ctps and eog scoring with xcorr.